### PR TITLE
Update to ffmpeg 6.0

### PR DESCRIPTION
**Pull request by wader**
Removed and renamed constants with entry from API changes https://git.ffmpeg.org/gitweb/ffmpeg.git/blob/HEAD:/doc/APIchanges

Remove use of AVCodecContext.sub_text_format
  2021-09-20 - 176b8d785bf - lavc 59.9.100 - avcodec.h
    Deprecate AVCodecContext.sub_text_format and the corresponding
    AVOptions. It is unused since the last major bump.

AV_CODEC_CAP_TRUNCATED removed:
  2021-09-20 - dd846bc4a91 - lavc 59.8.100 - avcodec.h codec.h
    Deprecate AV_CODEC_FLAG_TRUNCATED and AV_CODEC_CAP_TRUNCATED,
    as they are redundant with parsers.

AV_CODEC_CAP_AUTO_THREADS renamed to AV_CODEC_CAP_AUTO_THREADS
  2021-03-16 - 7d09579190 - lavc 58.132.100 - codec.h
    Add AV_CODEC_CAP_OTHER_THREADS as a new name for
    AV_CODEC_CAP_AUTO_THREADS. AV_CODEC_CAP_AUTO_THREADS
    is now deprecated.

AV_CODEC_CAP_INTRA_ONLY removed (use AV_CODEC_PROP_INTRA_ONLY instead): AV_CODEC_CAP_LOSSLESS removed (use AV_CODEC_PROP_LOESSNES instead):
  2020-05-21 - 13b1bbff0b - lavc 58.86.101 - avcodec.h
    Deprecated AV_CODEC_CAP_INTRA_ONLY and AV_CODEC_CAP_LOSSLESS.

AV_CODEC_FLAG_TRUNCATED removed:
AV_CODEC_CAP_TRUNCATED removed:
  2021-09-20 - dd846bc4a91 - lavc 59.8.100 - avcodec.h codec.h
    Deprecate AV_CODEC_FLAG_TRUNCATED and AV_CODEC_CAP_TRUNCATED,
    as they are redundant with parsers.

AV_CODEC_FLAG2_DROP_FRAME_TIMECODE removed:
  Not API changelog but was removed in f843460eb790d37e444e5946628f228421916537:
  avcodec/avcodec: Remove AV_CODEC_FLAG2_DROP_FRAME_TIMECODE

  It has been deprecated in 94d68a4
  and can't be set via AVOptions. The only codecs that use it
  (the MPEG-1/2 encoders) have private options for this.
  So remove it.

AVFMT_FLAG_PRIV_OPT removed:
  2021-03-03 - 2ff40b98ec - lavf 58.70.100 - avformat.h
    Deprecate AVFMT_FLAG_PRIV_OPT. It will do nothing
    as soon as av_demuxer_open() is removed.

Related to #1106

### CODE REVIEW

File: av/codec/codec.pyx
Comment (wader) [inline av/codec/codec.pyx:104]: Should possible by renamed to OTHER_THREADS now?

File: scripts/activate.sh
Comment (wader) [inline scripts/activate.sh:17]: Not sure when this is usually updated

**Comment by wader** [comment]
Also haven't yet looked close at API things that have been added since 5.x

**Comment by hmaarrfk** [comment]
Coming here looking for a similar patch.

Is there a way to conditionally apply the changes based on the detected version of libav at compile time? That way compatibility with FFMPEG 5 and 6 could both be maintained for some transition period.

**Comment by wader** [comment]
@hmaarrfk I would guess it's possible somehow but i'm a cython beginner so this PR was mostly based on what was already there. Also not sure if pyav has some policy about what verisons to support?

**Comment by hmaarrfk** [comment]
Remind me to look at how h5py does it.

**Comment by wader** [comment]
Me? 😄 I here by remind you to look how h5py does it 😬

Jokes aside, feel free to continue on my PR in a new PR etc

**Comment by hmaarrfk** [comment]
h5py uses the setup.py to create a file called `config.pxi` which defines a few constants related to the build time library:

https://github.com/h5py/h5py/blob/master/setup_build.py#L156

it then uses these constants in `IF` statements
https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#compile-time-definitions

**Comment by hmaarrfk** [comment]
hmm, im' having trouble with the Cython defined constants. I'm not sure how much I care given everything as a whole.

I'm going to try to see if you patch is still valid over at conda-forge to see if we can get the process moving.

**Comment by wader** [comment]
@hmaarrfk 👍 let me know if i can assist with something

**Comment by wader** [comment]
Sorry a bit confused, haven't used conda-forge before. So the new builds at https://anaconda.org/conda-forge/av/files are based on this branch? was manually triggered somehow?

**Comment by hmaarrfk** [comment]
> Sorry a bit confused, haven't used conda-forge before. So the new builds at https://anaconda.org/conda-forge/av/files are based on this branch? was manually triggered somehow?

Lets discuss conda(-forge) usage and any related questions in https://github.com/conda-forge/av-feedstock/issues/68

**Review by hmaarrfk — APPROVED** [review]
While this will loose support for ffmpeg 5, i think that this is somewhat a good move altogether.

**Comment by wader** [comment]
Is there any policy etc on which ffmpeg versions pyav supports? if ffmpeg 5 support need to be kept i might be able to spend some time on it next week at work, but i'm quite new to python bindings so any guidelines how to do would be great.

**Comment by hmaarrfk** [comment]
I can't press merge, currently just an other open source contributor eager to see this "officially in". lets wait until somebody with "merge authority" can chime in. they might just be busy...

**Comment by wader** [comment]
@hmaarrfk aha ok! same as me then :) yeap let's see

**Comment by badger200** [comment]
This fixed pyav for me on ffmpeg 6.0 macOS 10.9 MacPorts

**Comment by jlaine** [comment]
OK this is a start, thanks for this. However it is only a start, we also need to update our binary wheels to use FFmpeg 6.0, so I would suggest heading over to https://github.com/pyav-org/pyav-ffmpeg and submitting a PR there to build FFmpeg binaries for all our platforms.

**Comment by wader** [comment]
@jlaine Hey, ok! hopefully i can get some time to look at that. Should we also update versions of all the libraries at the same time?

**Comment by wader** [comment]
@jlaine I noticed you already started doing some CI fixes. Guess best to wait with ffmpeg 6 update until other CI things are done?

**Comment by daveisfera** [comment]
There's a [fork with this up and running](https://github.com/WyattBlue/pyav). I'd recommend those changes be merged into this repo

**Comment by WyattBlue** [comment]
Let's just merge this, then try to fix tests.

File: av/codec/codec.pyx
Comment (wader) [reply av/codec/codec.pyx:104]: @WyattBlue just a reminder: i guess this const should be `OTHER_THREADS` to be in sync with libav

**Comment by jlaine** [comment]
I'm not too happy with this, we need to get pyav-ffmpeg built with Ffmpeg 6 and used in our CI to claim we're compatible with FFmpeg 6!

**Pull request merged**
